File: src/Intersection.cpp
#include "Intersection.h"

Intersection::Intersection(std::vector<std::string> directions) {
    for (auto &direction : directions) {
        approachList.emplace_back(direction);
    }
}

std::vector<RoadApproach> &Intersection::approaches() { return approachList; }

const std::vector<RoadApproach> &Intersection::approaches() const { return approachList; }

std::optional<std::reference_wrapper<RoadApproach>> Intersection::findApproach(const std::string &direction) {
    for (auto &approach : approachList) {
        if (approach.direction() == direction) {
            return approach;
        }
    }
    return std::nullopt;
}

std::vector<std::string> Intersection::approachIds() const {
    std::vector<std::string> ids;
    ids.reserve(approachList.size());
    for (const auto &approach : approachList) {
        ids.push_back(approach.direction());
    }
    return ids;
}

void Intersection::applyPhase(const SignalPhase &phase) {
    for (auto &approach : approachList) {
        const auto &states = phase.getStates();
        const auto match = states.find(approach.direction());
        if (match != states.end()) {
            approach.light().setState(match->second);
        } else {
            approach.light().setState(LightState::AllRed);
        }
    }
}

File: python/cross_roads_ai/intersection.py
from typing import List, Optional
from .road_approach import RoadApproach
from .light_state import LightState
from .signal_phase import SignalPhase

class Intersection:
    def __init__(self, approach_ids: List[str]):
        self._approaches = [RoadApproach(d) for d in approach_ids]

    def approaches(self) -> List[RoadApproach]:
        return list(self._approaches)

    def approachIds(self) -> List[str]:
        return [a.direction() for a in self._approaches]

    def findApproach(self, approach_id: str) -> Optional[RoadApproach]:
        for app in self._approaches:
            if app.direction() == approach_id:
                return app
        return None

    def applyPhase(self, phase: SignalPhase) -> None:
        states = phase.approachStates()
        for approach in self._approaches:
            approach.light().setState(states.get(approach.direction(), LightState.Red))
